File: backend/app/models/match.py
from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class Match(BaseModel):
    """
    Representa uma partida da Copa do Mundo.
    """

    model_config = ConfigDict(from_attributes=True)

    # ==========================================================
    # Identificação
    # ==========================================================

    id: Optional[int] = Field(
        default=None,
        description="Identificador único da partida."
    )

    # ==========================================================
    # Seleções
    # ==========================================================

    home_team: str = Field(
        ...,
        description="Nome da seleção mandante."
    )

    away_team: str = Field(
        ...,
        description="Nome da seleção visitante."
    )

    # ==========================================================
    # Estádio
    # ==========================================================

    stadium: str = Field(
        ...,
        description="Nome do estádio."
    )

    city: str = Field(
        ...,
        description="Cidade."
    )

    country: str = Field(
        ...,
        description="País."
    )

    # ==========================================================
    # Data
    # ==========================================================

    date: str = Field(
        ...,
        description="Data e horário da partida."
    )

    time: str = Field(
        ...,
        description="Horário da partida."
    )

    # ==========================================================
    # Competição
    # ==========================================================

    stage: str = Field(
        default="GROUP",
        description="Fase da competição."
    )

    group: Optional[str] = Field(
        default=None,
        max_length=1,
        description="Grupo da partida (A-H)."
    )

    round: int = Field(
        default=1,
        ge=1,
        description="Rodada da competição."
    )

    # ==========================================================
    # Resultado
    # ==========================================================

    home_score: Optional[int] = Field(
        default=None,
        ge=0,
        description="Gols da seleção mandante."
    )

    away_score: Optional[int] = Field(
        default=None,
        ge=0,
        description="Gols da seleção visitante."
    )

    finished: bool = Field(
        default=False,
        description="Indica se a partida foi finalizada."
    )

    # ==========================================================
    # Auditoria
    # ==========================================================

    created_at: datetime = Field(
        default_factory=datetime.utcnow,
        description="Data de criação do registro."
    )

    updated_at: Optional[datetime] = Field(
        default=None,
        description="Data da última atualização."
    )

    # ==========================================================
    # Utilidades
    # ==========================================================

    @property
    def has_result(self) -> bool:
        """
        Verifica se a partida possui um resultado registrado.
        """

        return (
            self.home_score is not None
            and self.away_score is not None
        )

    @property
    def is_group_stage(self) -> bool:
        """
        Indica se a partida pertence à fase de grupos.
        """     

        return self.stage == "GROUP"

    @property
    def winner(self) -> Optional[str]:
        """
        Retorna o ID da seleção vencedora.

        None = empate ou partida sem resultado.
        """

        if not self.has_result:
            return None

        if self.home_score > self.away_score:
            return self.home_team

        if self.away_score > self.home_score:
            return self.away_team

        return None

    @property
    def is_draw(self) -> bool:
        """
        Verifica se a partida terminou empatada.
        """

        if not self.has_result:
            return False

        return self.home_score == self.away_score

File: frontend/src/app/types/Match.ts
import type { Team } from "./Team";
import type { Stadium } from "./Stadium";

/* ==========================================================
 * Fases da Copa
 * ========================================================== */

export type MatchStage =
    | "GROUP"
    | "ROUND_OF_32"
    | "ROUND_OF_16"
    | "QUARTER_FINAL"
    | "SEMI_FINAL"
    | "THIRD_PLACE"
    | "FINAL";

/* ==========================================================
 * Partida
 * ========================================================== */

export interface Match {
    /**
     * Identificador da partida.
     */
    id: number;

    /**
     * Data da partida.
     */
    date: string;

    /**
     * Horário da partida.
     */
    time: string;

    /**
     * Grupo (A, B, C...)
     */
    group?: string;

    /**
     * Fase da competição.
     */
    stage: MatchStage;

    /**
     * Rodada.
     */
    round: number;

    /**
     * Resultado oficial.
     */
    home_score?: number;

    away_score?: number;

    qualified_team_id?: number;

    /**
     * Indica se a partida foi finalizada.
     */
    finished: boolean;

    /**
     * Dados completos.
     */
    home_team: Team;

    away_team: Team;

    stadium: Stadium;

    /**
     * Datas de auditoria.
     */
    created_at?: string;

    updated_at?: string;
}
/* ==========================================================
 * Criação
 * ========================================================== */

export interface CreateMatchRequest {

    home_team_id: number;

    away_team_id: number;

    stadium_id: number;

    date: string;

    time: string;

    stage: MatchStage;

    group?: string;

    round: number;

    home_score?: number;

    away_score?: number;

    qualified_team_id?: number;

    finished: boolean;

}

/* ==========================================================
 * Atualização
 * ========================================================== */

export interface UpdateMatchRequest {

    id: number;

    home_team_id: number;

    away_team_id: number;

    stadium_id: number;

    date: string;

    time: string;

    stage: MatchStage;

    group?: string;

    round: number;

    home_score?: number;

    away_score?: number;

    qualified_team_id?: number;

    finished: boolean;

}
/* ==========================================================
 * Resultado
 * ========================================================== */

export interface MatchResult {
    home_score: number;

    away_score: number;

    qualified_team_id?: number;

    finished: boolean;
}

/* ==========================================================
 * Filtros
 * ========================================================== */

export interface MatchFilters {
    stage?: MatchStage;

    group?: string;

    round?: number;

    stadium_id?: number;

    team_id?: number;

    finished?: boolean;
}